cluster-autoscaler/proposals/buffers.md

```diff
@@ -0,0 +1,567 @@
+
+```
+apiversion: autoscaling.x-k8s.io/v1alpha1
+kind: Buffer
+metadata:
+  name: testing-capacity
+  namespace: my-namespace
+spec:
+  targetRef:
+    apiVersion: apps/v1
+    kind: CustomComputeClass
+    name: my-custom-class
+  type: buffer.x-k8s.io/active-capacity
+  recreationStrategy: keep-empty
+  nodeClassBufferSpec:
+    totalCpu: 40
+    totalMem: 5120Mi
+```
+
+Note: in the example the total size would be split into smaller chunks with some
+default shape. If a non-standard shape is needed it can be further defined in
+the `spec.nodeClassBufferSpec.perChunkResources`
+
+## Notes/Constraints/Caveats (Optional)
+
+*   The Buffer implementation assumes that the pods run as a part of scale
+    subresource are homogeneous.
+*   The initial implementation would not have an option to specify
+    ResourceClaims: the DRA integration is still in alpha phase and there are
+    implementation tasks left in CA to make such integration possible. This is
+    planned as a part of a next milestone.
+
+## Risks and Mitigations
+
+*   The implementation of buffers will introduce additional “virtual” pods to be
+    handled by the autoscalers. In case the buffer chunks were very small the
+    number of processed pods may grow - we do not expect that buffers would
+    introduce more chunks than the number of pods that user may want to run.
+    Since these are only virtual they will introduce less overhead in the
+    cluster than balloon pods/deployments (no need to preempt etc).
+
+# Design Details
+
+## Proposed API
+
+```go
+type Buffer struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+
+    Spec   BufferSpec
+    Status BufferStatus
+}
+
+type BufferSpec struct {
+    // Optional pointer to an object that defines how to interpret this buffer.
+    // Depending on the type of referenced object one of the configurations below will be used:
+    //   - for a crd implementing scale subresource (like deployment, job, replicaset)
+    //     the ReplicaBasedBufferSpec will be used
+    //   - for GKE CCC or Karpenter NodePool the NodeClassBufferSpec will be used
+    //   - in case of no target object the user will be able to directly define the cluster capacity
+    //   - … - custom crd if the user runs their own buffer controller that can handle this target
+    TargetRef *TypedLocalObjectReference
+    Capacity  BufferCapacity
+
+    // Initially only active stand by meaning scaling up the cluster will be available.
+    // In the future the cloud providers will be able to offer their own buffer types.
+    Type string
+
+    // How the buffer should be maintained over time:
+    // - KeepEmpty - upscale the cluster to keep the empty buffer on top of the running workloads
+    // - KeepMinimum - upscale the cluster and let workloads fill in the buffer as needed
+    RecreationStrategy string
+}
+
+type BufferCapacity struct {
+    // One of the following must be set
+    // +optional
+    Replicas *ReplicaBufferCapacity
+    // +optional
```

Comment:
If you have a pod-centric headroom concept, why do you also need a replica based and nodeclass based concept?

Reply:
Is it right to think of `NodeClassBufferCapacity` as syntactic sugar? Everything you can express with it you could also express with `PodBufferCapacity`, it's just less/simpler config?

As a side-note "NodeClass" might be as bit confusing for Karpenter users, as it's an orthogonal concept to Karpenter's NodeClass concept (maps better to its "NodePool" concept). 

WDYT about calling this `Nodes` for parity with what you have w/ `Pods`. Each is node-centric and pod-centric buffering, respectively.

Reply:
> Maybe just Resources and the type called ResourceBufferCapacity? @towca - wdyt?

LGTM!